spec/index.adoc

```diff
@@ -438,6 +438,15 @@ If the canister has a https://webassembly.github.io/spec/core/binary/modules.htm
 +
 It is recommended for the canister to have a custom section called "icp:public candid:service", which contains the UTF-8 encoding of https://github.com/dfinity/candid/blob/master/spec/Candid.md#core-grammar[the Candid interface] for the canister.
 
+* `/canister/<canister_id>/property/<name>` (blob):
++
+The systems maintains for each canister a generic key-value store, called the “canister properties”. These properties are provided by canister aministrators and while the system does not interpret them other platform components (e.g. the HTTP gateway protocol or Internet Identity) may. In contrast, metadata is provided by canister developers and may be interpreted by the system.
```

Comment:
Metadata is also not interpreted by the system, and can be thought of as key-value store. So I am not sure the difference between metadata and property. The current spec only specifies Wasm custom section as metadata, but it can contain other metadata as well. System or custom properties, such as `controllers`, `module_hash` and `disable_ic0_app_service_worker` can all be part of the metadata. It would be nice if we have a unified way of controlling these metadata/property.

Reply:
We could simplify this proposal by enabling metadata to be set via create_canister and changed via update_settings in addition to being set via custom sections, but the current WasmMetadata proto for example is specific to WasmCustomSection.   We also have issues with what happens if the user attempts to overwrite something specified in the metadata via update_settings.  Then there is the question of what is the provenance of the canister developer vs canister administrator.   I agree that perhaps the distinction between whether or not something effects the "system" is not clear and perhaps should be removed at least from the spec.  However the distinction that metadata comes from the wasm binary and properties come from canister settings seems pretty clear to me.

Reply:
The current WasmMetadata distinguishes CUSTOM_SECTION_TYPE_PUBLIC from CUSTOM_SECTION_TYPE_PRIVATE.  Is this distinction currently in use?  Is this something that we should/should reflect in properties ?

Reply:
I think the distinction between metadata and properties, as they exist right now (including this PR) is rather clear to me as well:

 * Metadata is tied to a canister _module_. Before you install code, there is no metadata. You can change it only together with the code. And if you uninstall the code, it’s gone.
 * Properties are tied to the canister _instance_. They exist even when the canister is empty (no code installed), and can be modified independently of the code.

It’s a good thing that “metadata” has a clear focus here – it’s metadata about the code – and I think that is why we kept it separate from other canister settings back then.

The naming is always a bit artificial, and maybe it helps to sometimes say “canister code metadata” to clarify, but I see no urgent need to unify things here.

Reply:
We were using `custom_section` as the endpoint name initial, but @rossberg wants something more conceptual. Quoting from earlier PR:

> Not sure I'm too fond of the name /custom_section/ in this path, since that's more of an implementation detail than a > conceptual notion. Maybe one day we want to have other/additional ways to provide such data. Abstracting this a  little also allows to embed naming conventions and additional information in the physical section names.
> 
> How about /metadata/ instead?

I understand the current distinction, but we have four endpoints now: `controllers`, `module_hash`, `metadata` and `property`. It's a bit wired to single out `controllers` and `module_hash` as top-level endpoints. By the current definition, `controllers` will be inside `property` and `module_hash` will be inside `metadata`.

Reply:
We could rename WasmCustomSection to MetadataValue and CustomSectionType to MetadataValueType and add a CANISTER_SETTING_TYPE_XXX type, then we could use metadata but only allow canister settings to see/update those bits of metadata with CANISTER_SETTINGS_TYPE_XXXX.  However, it wouldn't be clear in the state tree whether this was coming from a custom section or from canister settings.

Reply:
We could use `metadata/module` to represent data from Wasm module, and `metadata/setting` for canister settings. 

> The current WasmMetadata distinguishes CUSTOM_SECTION_TYPE_PUBLIC from CUSTOM_SECTION_TYPE_PRIVATE. Is this distinction currently in use? Is this something that we should/should reflect in properties ?

We specify the custom section visibility in the Wasm, but it's not visible in the HTTP endpoint. Non-controller user gets a 403 when accessing a private custom section.

Reply:
> By the current definition, controllers will be inside property

I agreee (hence my suggestion to add more settings instead of introducing properties in addition to settings).

> and module_hash will be inside metadata.

There is a crucial difference: Metadata can be _arbitrarily_ set by the developer. It comes _with_ the code, but the developer has control.

The module has is _derived_ from the actual code, and it would be horrible if the developer or the admin could manually set it. Here it really is the _system_ making a statement. So while it shares one property (heh) with the metadata (only exists when the canister is not empty), it also differs in crucial ways from metadata.